### PR TITLE
Checkpoint removal 2

### PR DESCRIPTION
**Pull request by tmarkovich**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

With 1bn entities, dumping the embedding table and reloading it was taking as long as training on a single 10bn edge chunk. This PR reduces the amount of check pointing to save us that cost which drastically speeds up single-instance embedding.

## How Has This Been Tested (if it applies)

Verified that this worked at runtime

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by tmarkovich** [comment]
@adamlerer @lw this is super hacky at the moment. I'll add some config options, but I wanted to get your feedback on this first before I went too far down a rabbit hole.

File: torchbiggraph/train_cpu.py [hunk 1]
```diff
@@ -797,19 +852,13 @@ def _swap_partitioned_embeddings(
...
-                embs = holder.partitioned_embeddings.pop((entity, part))
-                optimizer = self.trainer.partitioned_optimizers.pop((entity, part))
-                self.checkpoint_manager.write(
-                    entity, part, embs.detach(), optimizer.state_dict()
-                )
+                self._write_single_embedding(holder, entity, part)
                 self.embedding_storage_freelist[entity].add(embs.storage())
                 io_bytes += embs.numel() * embs.element_size()  # ignore optim state
                 # these variables are holding large objects; let them be freed
                 del embs
```
Comment (adamlerer) [inline torchbiggraph/train_cpu.py:859]: How do these lines work if you don't define `embs` and `optimizer` any more?

File: torchbiggraph/train_cpu.py [hunk 2]
```diff
@@ -770,11 +801,35 @@ def _load_embeddings(
...
+        )
+
+    def _nondestructive_write_embedding(self, bucket: Bucket):
+        parts: Set[Tuple[EntityName, Partition]] = set()
+        parts.update((e, bucket.lhs) for e in self.holder.lhs_partitioned_types)
+        parts.update((e, bucket.rhs) for e in self.holder.rhs_partitioned_types)
+        for entity, part in parts:
+            self._write_single_embedding(self.holder, entity, part)
+    
+    def _write_stats(self, bucket: Optional[Bucket], stats: Optional[BucketStats]):
```
Comment (adamlerer) [inline torchbiggraph/train_cpu.py:822]: This naming is misleading... I think it does more than write stats in the distributed scheduler.

**Review by adamlerer — COMMENTED** [review]
This change seems conceptually good to me. Of course, the code looks incorrect for the distributed case, but I assume you will fix that up for the final version.

We will probably need to test this change with some distributed runs. Unless you can make it very clearly correct "by inspection".

**Comment by tmarkovich** [comment]
> This change seems conceptually good to me. Of course, the code looks incorrect for the distributed case, but I assume you will fix that up for the final version.
> 
> We will probably need to test this change with some distributed runs. Unless you can make it very clearly correct "by inspection".

Even if it's "clearly correct by inspection", I'd still feel better if you all were able to run some distributed versions as well.
